### ISSUE TITLE
In-flight pause

### ISSUE THREAD

**Issue by Zarkonnen** [endless-sky/endless-sky#10186]
### Problem Description

For accessibility purposes, it would be very useful to be able to pause the game in-flight.

### Related Issue Links

* Prompted by [this Discord discussion](https://discord.com/channels/251118043411775489/308902312741568512/1249589276853669918).
* [Issue 4818](https://github.com/endless-sky/endless-sky/issues/4818) is about adjustable speed, and the conversation briefly touches upon pause.
* [This Steam forum discussion](https://steamcommunity.com/app/404410/discussions/2/1471966894880432300/) also suggests pause or slow mode.

### Desired Solution

Pressing the P key pauses the game in-flight. All controls for selection, giving orders, etc. are still available but time does not advance. A pause icon (⏸️) appears in the top left where the fast forward icon appears in caps lock fast mode.

### Alternative Approaches

A "near-pause" mode where game speed is reduced to 5% would give most of the same benefits and might be easier to implement if the game has trouble with a game speed multiplier of 0.

### Additional Context

![pause ui mockup](https://github.com/endless-sky/endless-sky/assets/898230/6daee29b-1f73-4af7-a66c-5dea60ef9cf6)

**Comment by Zarkonnen** [comment]
If there are no objections to the feature as suggested, I will start work on a pull request.

**Comment by TomGoodIdea** [comment]
There is a pause (under the <kbd>`</kbd> key), but it's currently only available if you launch Endless Sky in debug mode (`-d` flag).

**Comment by Zarkonnen** [comment]
Yes, exactly. Being able to see the game normally and interact with it, just with time not advancing.

**Comment by Zarkonnen** [comment]
Having had a brief look at the ` key pause functionality, it doesn't allow the player to interact with the game. For example, pausing the game like that and clicking on a ship does not select it.

I will look into alternate ways of implementing pause.

**Comment by Zarkonnen** [comment]
I'm looking at the code to see how pause could be implemented. I will leave notes in this comment for me or another future implementer.

The core problems with pause and arbitrary game speeds:
* the game runs on fixed-length ticks, and not delta time
* processing player input and advancing game time are not separated at the top level

As a result, the debug pause, which just pauses the main game loop entirely, makes it also stop processing inputs, so you can't e.g select ships when paused.

To disentangle this, Engine::Step and Engine::CalculateStep and the functions they call have to be modified to understand the idea of processing input only without advancing time. This could be handled by adding a `bool paused` field in PlayerInfo, which is then consulted by the code to check if time needs advancing.

But this doesn't get us the ability to arbitrarily change the game speed. Which would be a big accessibility benefit. This needs a multiplier of how much time should be advanced. By itself, this doesn't quite work though, because time quantities like e.g grudgeTime are measured in ticks.

I can think of two fixes to the underlying problem, both a bit painful in their own way:

1. Store two fields, `float timeScale` and `bool doTimeStep`, in PlayerInfo. Then, code that advances time in a way that's scalable - such as ship movement - can use timeScale, and code that is discrete, such as grudgeTime, advances if doTimeStep is true. The calling code then accumulates partial ticks and issues a doTimeStep = true every time a whole tick has occurred.
2. Rewrite all the time-based code to work on delta time instead of fixed steps. We still store `timeScale` in PlayerInfo but don't need `doTimeStep` because all the code that used discrete steps is rewritten to use delta time instead. So for example, grudgeTime turns from an int to a float.

Both of these are a major refactoring, and will require careful testing to make sure they don't introduce bugs, strange behaviours, or significant differences in play between normal speed and slowed down gameplay. (Such as AIs making decisions too quickly relative to time passing.)

**Comment by TheGiraffe3** [comment]
If you could make the pause functionality, it also might help with #7896. Being able to pause as nice is mentioned at the bottom of [this.](https://github.com/endless-sky/endless-sky/issues/7896#issuecomment-1356497085)

**Comment by Zarkonnen** [comment]
I asked tibetiroka on Discord if a refactor to delta time would be acceptable, and [they said no](https://discord.com/channels/251118043411775489/285540320823869441/1250361565648130058), for much the same reasons as outlined above.

I will instead concentrate on basic pause functionality, and maybe a half speed feature that reduces FPS to 30.

**Comment by Zarkonnen** [comment]
Half speed feature PR: https://github.com/endless-sky/endless-sky/pull/10207

**Comment by TheGiraffe3** [comment]
Zarkonnen has been somewhat inactive recently, and so the PR has unaddressed review comments. And because we're on a stable release now (0.10.9 just came out), it's unlikely that it'll be merged for another month or so.
